src/join/joining.cu

```diff
@@ -307,16 +309,199 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
 
 }
 
-gdf_error gdf_left_join(int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-                        gdf_column *left_result, gdf_column *right_result,
-                        gdf_context *join_context)
-{
-  return join_call<JoinType::LEFT_JOIN>(num_cols, leftcol, rightcol, left_result, right_result, join_context);
+size_t enumSize(gdf_dtype type) {
+    switch (type) {
+        case GDF_INT8:      return sizeof(int8_t);
+        case GDF_INT16:     return sizeof(int16_t);
+        case GDF_INT32:     return sizeof(int32_t);
+        case GDF_INT64:     return sizeof(int64_t);
+        case GDF_FLOAT32:   return sizeof(float);
+        case GDF_FLOAT64:   return sizeof(double);
+        case GDF_DATE32:    return sizeof(int32_t);
+        case GDF_DATE64:    return sizeof(int64_t);
+        case GDF_TIMESTAMP: return sizeof(int64_t);
+        default:            return 0;
+    }
 }
 
-gdf_error gdf_inner_join(int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-                         gdf_column *left_result, gdf_column *right_result,
-                         gdf_context *join_context)
-{
-  return join_call<JoinType::INNER_JOIN>(num_cols, leftcol, rightcol, left_result, right_result, join_context);
+template <JoinType join_type, typename size_type>
+gdf_error join_call_compute_df(
+                         gdf_column **left_cols, 
+                         int num_left_cols,
+                         int left_join_cols[],
+                         gdf_column **right_cols,
+                         int num_right_cols,
+                         int right_join_cols[],
+                         int num_cols_to_join,
+                         int result_num_cols,
+                         gdf_column **result_cols,
+                         gdf_column * left_indices,
+                         gdf_column * right_indices,
+                         gdf_context *join_context) {
+    //return error if the inputs are invalid
+    if ((left_cols == nullptr)  ||
+        (right_cols == nullptr)) { return GDF_DATASET_EMPTY; }
+
+    //check if combined join output is expected
+    bool compute_df = (result_cols != nullptr);
+
+    //return error if no output pointers are valid
+    if ( ((left_indices == nullptr)||(right_indices == nullptr)) &&
+         (!compute_df) ) { return GDF_DATASET_EMPTY; }
+
+    //If index outputs are not requested, create columns to store them
+    //for computing combined join output
+    gdf_column * left_index_out = left_indices;
+    gdf_column * right_index_out = right_indices;
+    std::vector<std::unique_ptr<gdf_column>> temporary_gdf_columns;
```

Comment:
I don't understand what the purpose of the `temporary_gdf_columns` is. If the passed in `left_indices` & `right_indices` are null, then just create them.
e.g.,
````
if(nullptr == left_indices){
   std::unique_ptr<gdf_column> left_col(new gdf_column);
   left_index_out == left_col.get();
}